src/NovelRT/Experimental/Graphics/GraphicsMemoryBlockCollection.cpp

```diff
@@ -0,0 +1,481 @@
+        return TryAllocate(size, 1, GraphicsMemoryRegionAllocationFlags::None, regions);
+    }
+
+    bool GraphicsMemoryBlockCollection::TrySetMinimumSize(size_t minimumSize)
+    {
+        std::lock_guard<std::mutex> guard(_mutex); // TODO: come back to this when Tanner gets back to you
+
+        size_t currentMinimumSize = GetMinimumSize();
+
+        if (minimumSize == currentMinimumSize)
+        {
+            return true;
+        }
+
+        size_t size = GetSize();
+        size_t blockCount = _blocks.size();
+
+        if (minimumSize < currentMinimumSize)
+        {
+            std::shared_ptr<GraphicsMemoryBlock> emptyBlock = nullptr;
+            size_t minimumBlockCount = GetMinimumBlockCount();
+
+            for (size_t blockIndex = blockCount; blockIndex-- != 0;)
+            {
+                std::shared_ptr<GraphicsMemoryBlock> block = _blocks[blockIndex];
+
+                size_t blockSize = block->GetSize();
+                bool blockIsEmpty = block->GetIsEmpty();
+
+                if (blockIsEmpty)
+                {
+                    if (((size - blockSize) >= minimumSize) && ((blockCount - 1) >= minimumBlockCount))
+                    {
+                        RemoveBlockAt(blockIndex);
+                        size -= blockSize;
+                        --blockCount;
+                    }
+                    else
+                    {
+                        emptyBlock = (emptyBlock == nullptr) ? block : emptyBlock;
+                    }
+                }
+            }
+            _emptyBlock = emptyBlock;
+        }
+        else
+        {
+            std::shared_ptr<GraphicsMemoryBlock> emptyBlock = nullptr;
+            size_t maximumBlockCount = GetMaximumBlockCount();
+            size_t maximumSharedBlockSize = GetMaximumSharedBlockSize();
+            size_t minimumBlockSize = GetMinimumBlockSize();
+
+            while (size < minimumSize)
+            {
+                if (blockCount < maximumBlockCount)
+                {
+                    size_t blockSize = GetAdjustedBlockSize(maximumSharedBlockSize);
+
+                    if (((size + blockSize) > minimumSize) && (blockSize != minimumBlockSize))
+                    {
+                        blockSize = minimumSize - size;
+                    }
+
+                    emptyBlock = (emptyBlock == nullptr) ? AddBlock(blockSize) : emptyBlock;
+                    size += blockSize;
+                }
+                else
+                {
+                    _emptyBlock = (_emptyBlock == nullptr) ? emptyBlock : _emptyBlock;
+                    return false;
+                }
+            }
+
+            _emptyBlock = (_emptyBlock == nullptr) ? emptyBlock : _emptyBlock;
+        }
+
+        _minimumSize = minimumSize;
+        if (_size != size)
+        {
+            throw std::runtime_error("Sizes don't match!"); // TODO: Make this a real exception later maybe? or just
+            // TODO: put it in debug builds only?
```

Comment:
Probably should make a decision here. Typically I don't think we'd `ifdef` this to be in debug only unless this is a check we _shouldn't_ be doing in production builds.... not sure here, though.

Reply:
TerraFX ifdefs it, which is why I wasn't sure how we'd want to handle it, as the frame-by-frame validation gets expensive in a release build.

Reply:
most validation checks like this are for internal consistency
its basically inline "unit tests" because you don't want those internal details to leak and so you can't test them "normally" really

and you don't want those enabled in release, it would kill perf

Reply:
We can `#ifdef` here too then, no? I know there was another area that had "debug only" validation that would benefit from this too

Reply:
it would be safer to wrap it in `assert` as ifdef debugs are compiler specific last I checked.

Reply:
I've made #342 to remind myself to resolve others later, but I ported this one to remove the TODO for now.